### PR TITLE
Cellular affinity test

### PR DESCRIPTION
**Pull request by anton-vinogradov**

### CODE REVIEW

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularAffinityBackupFilter.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     * @param previouslySelected A list of primary/backup nodes already chosen for a partition.
+     *                           The primary is first.
+     */
+    @Override public boolean apply(ClusterNode candidate, List<ClusterNode> previouslySelected) {
+        for (ClusterNode node : previouslySelected)
+            return Objects.equals(candidate.attribute(attrName), node.attribute(attrName));
+
+        return true;
+    }
+}
```
Comment (ivandasch) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularAffinityBackupFilter.java:59]: Oops, missing newline

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py [hunk 2]
```diff
@@ -47,10 +47,14 @@ class IgniteAwareService(BackgroundThreadService):
...
-    def __init__(self, context, num_nodes, modules, client_mode, version, properties):
+    def __init__(self, context, num_nodes, modules, client_mode, version, properties, jvm_options):
         super(IgniteAwareService, self).__init__(context, num_nodes)
 
-        self.jvm_options = context.globals.get("jvm_opts", "")
+        global_jvm_options = context.globals.get("jvm_opts", "")
+
+        service_jvm_options = " ".join(map(lambda x: '-J' + x, jvm_options)) if jvm_options else ""
+
+        self.jvm_options = " ".join(filter(None, [global_jvm_options, service_jvm_options]))
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py:57]: Cool!

File: modules/ducktests/tests/ignitetest/tests/cellular_affinity_test.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    @cluster(num_nodes=NUM_NODES * 3 + 1)
+    @parametrize(version=str(DEV_BRANCH))
+    def test(self, version):
+        """
+        Test Cellular Affinity scenario (partition distribution).
+        """
+        ignite_version = IgniteVersion(version)
+
+        ignites1 = IgniteService(
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/cellular_affinity_test.py:88]: Smells like a copy-paste? I suppose that service start can be extracted in separate method (may be nested method is better?)

File: modules/ducktests/tests/ignitetest/tests/cellular_affinity_test.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+        checker = IgniteApplicationService(
+            self.test_context,
+            java_class_name="org.apache.ignite.internal.ducktest.tests.cellular_affinity_test.DistributionChecker",
+            params={"cacheName": CellularAffinity.CACHE_NAME,
+                    "attr": CellularAffinity.ATTRIBUTE,
+                    "nodesPerCell": self.NUM_NODES},
+            version=ignite_version)
+
+        checker.start()
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/cellular_affinity_test.py:123]: Test without assertions or meauserements looks a little bit weird. May be assert smth in checker log? (log_monitor and wait_until, I mean)

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularAffinityBackupFilter.java (hunk 1 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularAffinityBackupFilter.java]: Done

File: modules/ducktests/tests/ignitetest/tests/cellular_affinity_test.py (hunk 3 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/tests/cellular_affinity_test.py:88]: Done

**Pull request merged**